### PR TITLE
3D Animation gallery followups

### PR DESCRIPTION
**Pull request by rossbar**
Follow-up to #7025 

The main things are:
 - Splits the original example into 2 examples: one for view updating and a second for the random walk
 - Refactor the random walk example to use attribute setting instead redrawing everything each frame

What do you think @lobpcg ?

### CODE REVIEW

**Comment by lobpcg** [comment]
> Follow-up to #7025 
> 
> The main things are:
>  - Splits the original example into 2 examples: one for view updating and a second for the random walk

Sure, this is easy. 

>  - Refactor the random walk example to use attribute setting instead redrawing everything each frame
> 
> What do you think @lobpcg ?

I don't know how, and unsure if it is even possible.

**Comment by rossbar** [comment]
This is a PR, so these are the things I've already done - I was hoping you could take a look and make sure I didn't butcher the original intent of your examples too much! Sorry for the confusion :)

**Comment by lobpcg** [comment]
> This is a PR, so these are the things I've already done - I was hoping you could take a look and make sure I didn't butcher the original intent of your examples too much! Sorry for the confusion :)

Sorry for my confusion. I'll have a look tomorrow.

**Comment by lobpcg** [comment]
> Follow-up to #7025 

Looks better, thanks! A few suggestions:
- keep both references in the basic example;
- use 3d spectral rather than spring in both examples - I would like to see it popularized.

Do you think that a frame redraw is never needed i.e. can always be avoided?

For a new follow up PR I would like to see a new hopefully similar example but which utilities 
blit=True as in https://matplotlib.org/stable/api/animation_api.html

**Comment by lobpcg** [comment]
Finally, rather than just deleting my original module you should rename it and edit to become one of your examples, to make your changs trackable and comparable with my original.

File: examples/3d_drawing/plot_3d_animation_walk.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+def init():
+    ax.clear()
+    ax.scatter(*nodes.T, alpha=0.2, s=100, color="blue")
+    for vizedge in edges:
+        ax.plot(*vizedge.T, color="gray")
+    # Intialize the "walking" node
+    ax.plot(*nodes[0], alpha=1, marker="s", color="red")
+    ax.grid(False)
+    ax.set_axis_off()
+    plt.tight_layout()
```
Comment (lobpcg) [inline examples/3d_drawing/plot_3d_animation_walk.py:44]: This gives me a warning:
UserWarning: The figure layout has changed to tight
  plt.tight_layout()
  
delete, and use 
`fig = plt.figure(layout='tight')` 
below instead

File: examples/3d_drawing/plot_3d_animation_walk.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    if index % 5 == 0:
+        # Update node
+        node = random.choice(neighbors)
+        # Update the last line object, which corresponds to the walking node
+        pt = ax.lines[-1].set_data_3d(nodes[node][:, np.newaxis])
+    # Update view
+    ax.view_init(index * 0.2, index * 0.5)
+
+
+fig = plt.figure()
```
Comment (lobpcg) [inline examples/3d_drawing/plot_3d_animation_walk.py:60]: fig = plt.figure(layout='tight')

File: examples/3d_drawing/plot_3d_animation_walk.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+def _frame_update(index):
+    global node
+    # Update "current" node every 5 frames
+    neighbors = list(G.neighbors(node))
+    if index % 5 == 0:
+        # Update node
+        node = random.choice(neighbors)
+        # Update the last line object, which corresponds to the walking node
+        pt = ax.lines[-1].set_data_3d(nodes[node][:, np.newaxis])
```
Comment (lobpcg) [inline examples/3d_drawing/plot_3d_animation_walk.py:55]: `pt` is unused. Use `_` instead

File: examples/3d_drawing/plot_3d_animation_basic.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+"""
+===============
+Basic Animation
+===============
+
+NetworkX supports several 3D layout functions for visualization.
+These layouts can be combined with :external+matplotlib:doc:`api/animation_api`
+for 3D graph visualization.
+
+This example shows a basic animation incrementing the camera view.
```
Comment (lobpcg) [inline examples/3d_drawing/plot_3d_animation_basic.py:10]: Add from my original PR: 
```
Following 
https://sphinx-gallery.github.io/stable/auto_examples/plot_8_animations.html
using frame rotation of an initial plot of a graph as in 
https://matplotlib.org/stable/api/animation_api.html
```

File: examples/3d_drawing/plot_3d_animation_basic.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+# In this example, a frame update is only a rotation of a given 3D graph.
+
+
+def init():
+    ax.scatter(*nodes.T, alpha=0.2, s=100, color="blue")
+    for vizedge in edges:
+        ax.plot(*vizedge.T, color="gray")
+    ax.grid(False)
+    ax.set_axis_off()
+    plt.tight_layout()
```
Comment (lobpcg) [inline examples/3d_drawing/plot_3d_animation_basic.py:41]: remove. use instead below
fig = plt.figure(layout='tight')

File: examples/3d_drawing/plot_3d_animation_walk.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+from matplotlib import animation
+
+###############################################################################
+# Define a graph to plot.
+# -----------------------
+#
+# Pick a graph for visualization in 3D, such as one of the Platonic solids
+
+G = nx.dodecahedral_graph()
+pos = nx.spring_layout(G, dim=3, seed=2734)
```
Comment (lobpcg) [inline examples/3d_drawing/plot_3d_animation_walk.py:23]: pos = nx.spectral_layout(G, dim=3)

File: examples/3d_drawing/plot_3d_animation_basic.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+edges = np.array([(pos[u], pos[v]) for u, v in G.edges()])
+
+###############################################################################
+# Rotating 3D graph animation.
+# ----------------------------
+#
+# In this example, a frame update is only a rotation of a given 3D graph.
+
+
+def init():
```
Comment (lobpcg) [inline examples/3d_drawing/plot_3d_animation_basic.py:35]: do not you need here
```
ax.clear()
```
?

File: examples/3d_drawing/plot_3d_animation_walk.py (hunk 1 above)
Comment (rossbar) [reply examples/3d_drawing/plot_3d_animation_walk.py]: The problem with this is that the "tightness" can only be determined after the axis object is added. I went ahead and used `fig.tight_layout()` after the `fig.add_subplot` call, which should be equivalent.

File: examples/3d_drawing/plot_3d_animation_walk.py (hunk 2 above)
Comment (rossbar) [reply examples/3d_drawing/plot_3d_animation_walk.py:59]: See above

File: examples/3d_drawing/plot_3d_animation_basic.py (hunk 4 above)
Comment (rossbar) [reply examples/3d_drawing/plot_3d_animation_basic.py:10]: I removed this intentionally, on the idea that the networkx gallery should really focus on network visualization specifically. Matplotlib's and sphinx-gallery's examples and reference docs are a much better source for this info.

File: examples/3d_drawing/plot_3d_animation_basic.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+from matplotlib import animation
+
+###############################################################################
+# Define a graph to plot.
+# -----------------------
+#
+# Pick a graph for visualization in 3D, such as one of the Platonic solids
+
+G = nx.dodecahedral_graph()
+pos = nx.spring_layout(G, dim=3, seed=2734)
```
[the remark this answers is not in the document]
Comment (rossbar) [reply examples/3d_drawing/plot_3d_animation_basic.py:24]: It looks to me like there may be something wrong with the multi-dimensional spectral layout... when I try this locally, it results in a "flat" dodecahedron whereas the spring layout looks much better.

I'd vote to keep spring here and address the potential issue with spectral layout in a separate issue.

File: examples/3d_drawing/plot_3d_animation_basic.py (hunk 7 above)
Comment (rossbar) [reply examples/3d_drawing/plot_3d_animation_basic.py:35]: It's necessary to remove the initial point that is added to represent the current node. Without this, one of the vertices of the dodecahedron retains a permanent red square.

**Review by rossbar — COMMENTED** [review]
Despite the fact that this PR creates two examples from one (doubling the number of animations), it still improves the total example runtime (from 11sec -> ~6sec on my local machine) compared to the version on `main`.

**Comment by lobpcg** [comment]
> Finally, rather than just deleting my original module you should rename it and edit to become one of your examples, to make your changes trackable and comparable with my original.

Also, it may not be the best practice for a contributor to mark a suggestion as "resolved" unless possibly in trivial cases as it makes the follow-up reviewing more difficult...

**Comment by rossbar** [comment]
> Finally, rather than just deleting my original module you should rename it and edit to become one of your examples, to make your changs trackable and comparable with my original.

FWIW I had done so, i.e. used `git mv` to rename the file into one of the split examples. However, I am not sure how you can do this across "child" files, i.e. how to maintain provenance from one parent -> two different child files.

The only thing I could think to do to alleviate this was the following:
 1. Create a copy of the original example file (as-is, no content changes) to the new example file name, i.e. `cp plot_3d_rotation_animation.py plot_3d_animation_walk.py`
 2. Commit that file with the `--author` tag to preserve original attribution
 3. Then, `git mv` the original file to rename it: `git mv plot_3d_rotation_animation.py plot_3d_animation_basic.py`

At this point, there are two new files with content identical to the original `plot_3d_rotation_animation.py` file, and attribution (i.e. file authorship) is preserved for both new files.

Now, to make the changes to the file contents reviewable in github, I can squash all of the changes from the individual commits and apply them to the new files.

I couldn't figure out how to do this all with only rebasing, so I had to resort to creating a new branch and using a combination of `squash` and `restore` to map changes across branches.

**Comment by rossbar** [comment]
Closing in favor of #7681 . FWIW the changes to the content of the files are now viewable in the GitHub UI by inspecting commit fd27267

**Pull request closed without merging**